File: core/src/main/java/dev/bee/kanjianki/core/NoteTypeFieldMappingPolicy.java
package dev.bee.kanjianki.core;

import java.util.Collections;
import java.util.List;

public final class NoteTypeFieldMappingPolicy {
    private NoteTypeFieldMappingPolicy() {
    }

    public static FieldGuesses guessFields(List<String> fields) {
        List<String> safeFields = fields == null ? Collections.emptyList() : fields;
        RecordsSyncModels.Settings defaults = RecordsSyncModels.Settings.kikuDefaults();
        String expression = firstMatchingField(
                safeFields,
                defaults.expressionField,
                "Front",
                "Japanese",
                "Word",
                "Vocabulary",
                "Term"
        );
        String meaning = firstMatchingField(
                safeFields,
                defaults.meaningField,
                "Meaning",
                "Back",
                "Definition",
                "Glossary"
        );
        if (expression.trim().isEmpty() && !safeFields.isEmpty()) {
            expression = safeFields.get(0);
        }
        if (meaning.trim().isEmpty() && safeFields.size() > 1) {
            meaning = safeFields.get(1);
        }
        return new FieldGuesses(
                expression,
                firstMatchingField(safeFields, defaults.readingField, "Reading", "Kana", "Pronunciation"),
                meaning,
                firstMatchingField(safeFields, defaults.sentenceField, "Context", "Example", "ExampleSentence"),
                firstMatchingField(safeFields, defaults.frequencyField, "Freq"),
                firstMatchingField(safeFields, defaults.frequencySortField, "FrequencySort", defaults.frequencyField)
        );
    }

    public static String firstMatchingField(List<String> fields, String... candidates) {
        List<String> safeFields = fields == null ? Collections.emptyList() : fields;
        for (String candidate : candidates) {
            for (String field : safeFields) {
                if (field != null && candidate != null && field.equalsIgnoreCase(candidate)) {
                    return field;
                }
            }
        }
        return "";
    }

    public static final class FieldGuesses {
        public final String expression;
        public final String reading;
        public final String meaning;
        public final String sentence;
        public final String frequency;
        public final String frequencySort;

        private FieldGuesses(
                String expression,
                String reading,
                String meaning,
                String sentence,
                String frequency,
                String frequencySort
        ) {
            this.expression = expression == null ? "" : expression;
            this.reading = reading == null ? "" : reading;
            this.meaning = meaning == null ? "" : meaning;
            this.sentence = sentence == null ? "" : sentence;
            this.frequency = frequency == null ? "" : frequency;
            this.frequencySort = frequencySort == null ? "" : frequencySort;
        }
    }
}

File: core/src/main/kotlin/dev/bee/kanjianki/core/NoteTypeFieldMappingPolicy.kt
package dev.bee.kanjianki.core

import java.util.Collections
import kotlin.jvm.JvmWildcard

object NoteTypeFieldMappingPolicy {
    @JvmStatic
    fun guessFields(fields: List<String?>?): FieldGuesses {
        val safeFields = fields.orEmpty()
        val defaults = RecordsSyncModels.Settings.kikuDefaults()
        var expression = firstMatchingField(
            safeFields,
            defaults.expressionField,
            "Front",
            "Japanese",
            "Word",
            "Vocabulary",
            "Term",
        )
        var meaning = firstMatchingField(
            safeFields,
            defaults.meaningField,
            "Meaning",
            "Back",
            "Definition",
            "Glossary",
        )
        if (expression.trim().isEmpty() && safeFields.isNotEmpty()) {
            expression = safeFields[0].orEmpty()
        }
        if (meaning.trim().isEmpty() && safeFields.size > 1) {
            meaning = safeFields[1].orEmpty()
        }
        return FieldGuesses.create(
            expression,
            firstMatchingField(safeFields, defaults.readingField, "Reading", "Kana", "Pronunciation"),
            meaning,
            firstMatchingField(safeFields, defaults.sentenceField, "Context", "Example", "ExampleSentence"),
            firstMatchingField(safeFields, defaults.frequencyField, "Freq"),
            firstMatchingField(safeFields, defaults.frequencySortField, "FrequencySort", defaults.frequencyField),
        )
    }

    @JvmStatic
    fun firstMatchingField(fields: List<String?>?, vararg candidates: String?): String {
        val safeFields = fields.orEmpty()
        for (candidate in candidates) {
            for (field in safeFields) {
                if (field != null && candidate != null && field.equals(candidate, ignoreCase = true)) {
                    return field
                }
            }
        }
        return ""
    }

    @JvmStatic
    fun choice(name: String?, fields: List<String?>?): NoteTypeChoice {
        return NoteTypeChoice(name, fields)
    }

    @JvmStatic
    fun label(noteType: NoteTypeChoice?): String {
        val safeNoteType = noteType ?: choice("", emptyList())
        return safeNoteType.name() + " (" + StudyTextCopy.countText(safeNoteType.fields().size, "field", "fields") + ")"
    }

    @JvmStatic
    fun labels(noteTypes: List<@JvmWildcard NoteTypeChoice?>?): Array<String> {
        return noteTypes.orEmpty().map { label(it) }.toTypedArray()
    }

    class NoteTypeChoice(name: String?, fields: List<String?>?) {
        private val nameValue = name.orEmpty()
        private val fieldsValue: List<String>

        init {
            @Suppress("UNCHECKED_CAST")
            fieldsValue = Collections.unmodifiableList(ArrayList(fields.orEmpty())) as List<String>
        }

        fun name(): String {
            return nameValue
        }

        fun fields(): List<String> {
            return fieldsValue
        }
    }

    class FieldGuesses private constructor(
        @JvmField val expression: String,
        @JvmField val reading: String,
        @JvmField val meaning: String,
        @JvmField val sentence: String,
        @JvmField val frequency: String,
        @JvmField val frequencySort: String,
    ) {
        companion object {
            @JvmSynthetic
            fun create(
                expression: String?,
                reading: String?,
                meaning: String?,
                sentence: String?,
                frequency: String?,
                frequencySort: String?,
            ): FieldGuesses {
                return FieldGuesses(
                    expression.orEmpty(),
                    reading.orEmpty(),
                    meaning.orEmpty(),
                    sentence.orEmpty(),
                    frequency.orEmpty(),
                    frequencySort.orEmpty(),
                )
            }
        }
    }
}
